src/Microsoft.AspNetCore.Routing/Template/TemplateBinder.cs

```diff
@@ -58,7 +58,47 @@ public class TemplateBinder
         // Step 1: Get the list of values we're going to try to use to match and generate this URI
         public TemplateValuesResult GetValues(RouteValueDictionary ambientValues, RouteValueDictionary values)
         {
-            var context = new TemplateBindingContext(_defaults);
+            // Any default values that don't appear as parameters are treated like filters. Any new values
+            // provided must match these defaults.
+            foreach (var filter in _filters)
+            {
+                var parameter = GetParameter(filter.Key);
+                if (parameter != null)
+                {
+                    continue;
+                }
+
+                object value;
+                if (values.TryGetValue(filter.Key, out value))
+                {
+                    if (!RoutePartsEqual(value, filter.Value))
+                    {
+                        // If there is a non-parameterized value in the route and there is a
+                        // new value for it and it doesn't match, this route won't match.
+                        return null;
+                    }
+                }
+            }
+
+            // Validate that all required parameters have a value.
+            // Perf: Fail fast if the required parameter doesn't have a value.
+            for (var i = 0; i < _template.Parameters.Count; i++)
+            {
+                var parameter = _template.Parameters[i];
+                if (parameter.IsOptional || parameter.IsCatchAll)
+                {
+                    continue;
+                }
+
+                if (!HasValue(parameter.Name, ambientValues, values))
```

Comment:
Basically the process here is:
1. Generate accepted values
2. Ensure the accepted values can generate a URL (filters, required parameters)
3. Run constraints
4. Generate URL

You're trying to optimize this code by doing 2 before doing 1, and the problem is that 1 is a fairly complex process. 

Perhaps what you could do is break out 1 into it's own method, with an emphasis on no-op-ing in the common case where the `values` RVD doesn't need to change. Then make sure the rest of the process treats accepted values as readonly.